examples/tictactoe.rs

```diff
@@ -0,0 +1,297 @@
+
+#[derive(Debug)]
+pub enum Error {
+	OutOfRange,
+	AlreadyOccupied,
+	Interpreter(InterpreterError),
+	Deserialize(DeserializationError),
+	Validation(ValidationError),
+}
+
+impl fmt::Display for Error {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		write!(f, "{:?}", self)
+	}
+}
+
+impl From<InterpreterError> for Error {
+	fn from(e: InterpreterError) -> Self {
+		Error::Interpreter(e)
+	}
+}
+
+impl From<DeserializationError> for Error {
+	fn from(e: DeserializationError) -> Error {
+		Error::Deserialize(e)
+	}
+}
+
+impl From<ValidationError> for Error {
+	fn from(e: ValidationError) -> Error {
+		Error::Validation(e)
+	}
+}
+
+impl HostError for Error {}
+
+mod tictactoe {
+	use super::Error;
+
+	#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+	pub enum Player {
+		X,
+		O,
+	}
+
+	#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+	pub enum GameResult {
+		Draw,
+		Won(Player),
+	}
+
+	impl Player {
+		pub fn into_i32(maybe_player: Option<Player>) -> i32 {
+			match maybe_player {
+				None => 0,
+				Some(Player::X) => 1,
+				Some(Player::O) => 2,
+			}
+		}
+	}
+
+	#[derive(Debug)]
+	pub struct Game {
+		board: [Option<Player>; 9],
+	}
+
+	impl Game {
+		pub fn new() -> Game {
+			Game {
+				board: [None; 9],
+			}
+		}
+
+		pub fn set(&mut self, idx: i32, player: Player) -> Result<(), Error> {
+			if idx < 0 || idx > 9 {
+				return Err(Error::OutOfRange);
+			}
+			if self.board[idx as usize] != None {
+				return Err(Error::AlreadyOccupied);
+			}
+			self.board[idx as usize] = Some(player);
+			Ok(())
+		}
+
+		pub fn get(&self, idx: i32) -> Result<Option<Player>, Error> {
+			if idx < 0 || idx > 9 {
+				return Err(Error::OutOfRange);
+			}
+			Ok(self.board[idx as usize])
+		}
+
+		pub fn game_result(&self) -> Option<GameResult> {
+			// 0, 1, 2
+			// 3, 4, 5
+			// 6, 7, 8
+			let patterns = &[
+				// Rows
+				(0, 1, 2),
+				(3, 4, 5),
+				(6, 7, 8),
+
+				// Columns
+				(0, 3, 6),
+				(1, 4, 7),
+				(2, 5, 8),
+
+				// Diagonals
+				(0, 4, 8),
+				(2, 4, 6),
+			];
+
+			// Returns Some(player) if all cells contain same Player.
+			let all_same = |i1: usize, i2: usize, i3: usize| -> Option<Player> {
+				if self.board[i1].is_none() {
+					return None;
+				}
+				if self.board[i1] == self.board[i2] && self.board[i2] == self.board[i3] {
+					return self.board[i1];
+				}
+				None
+			};
+
+			for &(i1, i2, i3) in patterns {
+				if let Some(player) = all_same(i1, i2, i3) {
+					return Some(GameResult::Won(player));
+				}
+			}
+
+			// Ok, there is no winner. Check if it's draw.
+			let all_occupied = self.board.iter().all(|&cell| cell.is_some());
+			if all_occupied {
+				Some(GameResult::Draw)
+			} else {
+				// Nah, there are still empty cells left.
+				None
+			}
+		}
+	}
+}
+
+struct Runtime<'a> {
+	player: tictactoe::Player,
+	game: &'a mut tictactoe::Game,
+}
+
+const SET_FUNC_INDEX: usize = 0;
+const GET_FUNC_INDEX: usize = 1;
+
+impl<'a> Externals for Runtime<'a> {
```

Comment:
Some kind of macro (which I think @gavofyork put together in the Polkadot repo) for writing type-safe externals from regular Rust code would be a good wishlist item.

Reply:
definitely! This macro is planed as a second stage (a.k.a high level api)